File: src/main/java/org/apache/sysml/hops/spoof2/rewrite/RewriteDistributiveSumProduct.java
package org.apache.sysml.hops.spoof2.rewrite;

import java.util.ArrayList;
import java.util.List;

import org.apache.sysml.hops.Hop.AggOp;
import org.apache.sysml.hops.spoof2.plan.SNode;
import org.apache.sysml.hops.spoof2.plan.SNodeAggregate;
import org.apache.sysml.hops.spoof2.plan.SNodeNary;
import org.apache.sysml.hops.spoof2.plan.SNodeNary.NaryOp;

public class RewriteDistributiveSumProduct extends SPlanRewriteRule
{
	@Override
	public SNode rewriteNode(SNode parent, SNode node, int pos) {
		
		//pattern: agg(sum)-b(*)
		if( node instanceof SNodeAggregate && ((SNodeAggregate)node).getOp()==AggOp.SUM 
			&& node.getInput().get(0) instanceof SNodeNary 
			&& ((SNodeNary)node.getInput().get(0)).getOp()==NaryOp.MULT )
		{
			SNodeAggregate sum = (SNodeAggregate) node;
			SNodeNary mult = (SNodeNary) node.getInput().get(0);
			
			//check left/right aggregation pushdown
			int numApplied = 0;
			for( int i=0; i<2; i++ ) {
				SNode input = mult.getInput().get(i);
				if( !input.isScalar() ) {
					List<String> tgtSchema = new ArrayList<String>();
					List<Long> tgtDims = new ArrayList<Long>();
					//construct minimal target scheme for binary input
					for( String attr : input.getSchema() )
						if( sum.getSchema().contains(attr)
							|| mult.containsJoinCondition(attr) ) {
							tgtSchema.add(attr);
							tgtDims.add(input.getDim(attr));
						}
					//create new aggregate if potential for pre-agg
					if( input.getSchema().size() > tgtSchema.size() ) {
						SNodeAggregate sum2 = new SNodeAggregate(input, AggOp.SUM, tgtSchema);
						sum2.setDims(tgtDims.toArray(new Long[0]));
						SNodeRewriteUtils.replaceChildReference(mult, input, sum2);
						//update multiply dimensions and schema
						//TODO extend for row/column sums
						if( tgtSchema.isEmpty() ) { 
							mult.setDims(0L, 0L);
							mult.setSchema(mult.getInput().get(i).getSchema());
						}
						numApplied ++;
					}
				}
			}
			
			if( numApplied > 0 )
				LOG.debug("Applied RewriteDistributiveSumProduct (num="+numApplied+").");
		}
		
		return node;
	}

}

File: src/main/java/org/apache/sysml/hops/spoof2/rewrite/RewriteDistributiveSumProduct.kt
package org.apache.sysml.hops.spoof2.rewrite

import org.apache.sysml.hops.Hop.AggOp
import org.apache.sysml.hops.spoof2.plan.SNode
import org.apache.sysml.hops.spoof2.plan.SNodeAggregate
import org.apache.sysml.hops.spoof2.plan.SNodeNary
import org.apache.sysml.hops.spoof2.plan.SNodeNary.NaryOp

class RewriteDistributiveSumProduct : SPlanRewriteRule() {
    override fun rewriteNode(parent: SNode, node: SNode, pos: Int): SNode {

        //pattern: agg(sum)-b(*)
        if (node is SNodeAggregate && node.op == AggOp.SUM
                && node.inputs[0] is SNodeNary
                && (node.inputs[0] as SNodeNary).op == NaryOp.MULT) { // todo generalize to other multiply functions that are semiring to +
            val agg = node
            val mult = node.inputs[0] as SNodeNary
            // the attributes that occur in more than one input
            val joinAttrs = mult.getJoinLabelCounts().entrySet().filter { it.count > 1 }.map { it.element!! }

            //check left/right aggregation pushdown
            var numApplied = 0
            for (i in mult.inputs.indices) {
                val input = mult.inputs[i]
                // find indices that are neither in the output nor in the join condition
                val preAggAttrs = input.schema.labels.filter { it !in agg.schema && it !in joinAttrs }
                if (preAggAttrs.isNotEmpty()) {
                    // pre-aggregate these indices!
                    val preAgg = SNodeAggregate(AggOp.SUM, input, preAggAttrs.toHashSet())
                    preAgg.updateSchema() // todo - automate this?
                    SNodeRewriteUtils.replaceChildReference(mult, input, preAgg)
                    mult.updateSchema()

                    numApplied++
                }
            }

            if (numApplied > 0)
                SPlanRewriteRule.LOG.debug("Applied RewriteDistributiveSumProduct (num=$numApplied).")
        }

        return node
    }

}
